File: LeetCode/Medium/0983_Minimum_Cost_for_Tickets/Solution.cpp
// 2023-03-28 (tabulation)
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        /*
        notes:
        - days is a list of values from 1 to 365
        - tickets can either be bought for 1, 7, or 30 days
        - want to minimize the total $$ spent to reach the last day
        - don't we just iterate until we reach the thresholds?
        - this is definitely DP so we need to have n slots for our tabulation
        - just iterate until we hit thresholds for 1/7/30 for each day slot?

        analysis:
        - time = O(n) where n=length of days (though technically max is 365*30 so we can say O(1))
        - space = O(n) where n=length of days (though technically max is 365*30 so we can say O(1))
        */
        int n = days.size();
        // represents smallest cost to reach day i where day[i] = tab[i+1];
        vector<int> tab(n+1, INT_MAX); 
        tab[0] = 0;
        for (int i=0; i<n; i++) {
            for (int j=i+1; j<n+1 && days[j-1]-days[i] < 30; j++) {
                int dayDiff = days[j-1]-days[i];
                if (dayDiff < 1) // 1 day ticket
                    tab[j] = min(tab[j], tab[i]+costs[0]);
                if (dayDiff < 7) // 7 day ticket
                    tab[j] = min(tab[j], tab[i]+costs[1]);
                if (dayDiff < 30) // 30 day ticket
                    tab[j] = min(tab[j], tab[i]+costs[2]);
            }
        }

        return tab.back();
    }
};

File: LeetCode/Medium/0983_Minimum_Cost_for_Tickets/Solution.py
## 2021-11-25 (memoization)
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        cheapest = [-1] * len(days)
        return self.findCheapestPlan(days, costs, cheapest)
    
    def findCheapestPlan(self, days: List[int], costs: List[int], cheapest: List[int]) -> int:
        if len(days) == 0:
                return 0
        
        position = len(cheapest) - len(days)
        
        if cheapest[position] != -1:
            return cheapest[position]
        
        one_day_price = costs[0] + self.findCheapestPlan(days[1:], costs, cheapest)
        
        current_day = days[0]-1
        future_day = 0
        
        while future_day < len(days) and days[future_day] - current_day <= 7:
            future_day = future_day + 1
        seven_day_price = costs[1] + self.findCheapestPlan(days[future_day:], costs, cheapest)
        
        while future_day < len(days) and days[future_day] - current_day <= 30:
            future_day = future_day + 1
        thirty_day_price = costs[2] + self.findCheapestPlan(days[future_day:], costs, cheapest)
        
        cheapest[position] = min(one_day_price, seven_day_price, thirty_day_price)
        return cheapest[position]